prometheus/graphite/bridge.go

```diff
@@ -0,0 +1,230 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package graphite provides a bridge to push Prometheus metrics to a Graphite
+// server.
+package graphite
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"regexp"
+	"sort"
+	"strings"
+	"time"
+
+	"golang.org/x/net/context"
+
+	dto "github.com/prometheus/client_model/go"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/expfmt"
+	"github.com/prometheus/common/log"
+	"github.com/prometheus/common/model"
+)
+
+// Config defines the graphite bridge config.
+type Config struct {
+	// The url to push data to. Required.
+	URL string
+
+	// The interval to use for pushing data to Graphite. Defaults to 15 seconds.
+	Interval time.Duration
+
+	// The Gatherer to use for metrics. Defaults to prometheus.DefaultGatherer.
+	Gatherer prometheus.Gatherer
+
+	// The logger that messages are written to. Defaults to log.Base().
+	Logger log.Logger
+
+	// The prefix for your graphite metric. Defaults to empty string.
+	Prefix string
+}
+
+// Bridge pushes metrics to the configured graphite server.
+type Bridge struct {
+	url      string
+	interval time.Duration
+	prefix   string
+
+	g      prometheus.Gatherer
+	logger log.Logger
+}
+
+// NewBridge returns a pointer to a new Bridge struct.
+func NewBridge(c *Config) (*Bridge, error) {
+	b := &Bridge{}
+
+	if c.URL == "" {
+		return nil, errors.New("graphite bridge: no url given")
+	}
+	b.url = c.URL
+
+	if c.Gatherer == nil {
+		b.g = prometheus.DefaultGatherer
+	} else {
+		b.g = c.Gatherer
+	}
+
+	if c.Logger == nil {
+		b.logger = log.Base()
+	} else {
+		b.logger = c.Logger
+	}
+
+	if c.Prefix != "" {
+		b.prefix = c.Prefix
+	}
+
+	var z time.Duration
+	if c.Interval == z {
+		b.interval = 15 * time.Second
+	} else {
+		b.interval = c.Interval
+	}
+
+	return b, nil
+}
+
+// Run starts the event loop that pushes Prometheus metrics to Graphite at the
+// configured interval.
+func (b *Bridge) Run(ctx context.Context) {
+	ticker := time.NewTicker(b.interval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			if err := b.Push(); err != nil {
+				b.logger.Errorf("%v", err)
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// Push pushes Prometheus metrics to the configured Graphite server.
+func (b *Bridge) Push() error {
+	mfs, err := b.g.Gather()
+	if err != nil {
+		return err
+	}
+
+	now := time.Now().Unix()
+	// TODO: Write directly to conn?
+	buf, err := toReader(mfs, b.prefix, now)
+	if err != nil {
+		return err
+	}
+
+	// TODO: Should we expose a deadline to the user?
+	conn, err := net.Dial("tcp", b.url)
+	if err != nil {
+		return err
+	}
+	defer conn.Close()
+
+	_, err = io.Copy(conn, buf)
+	return err
+}
+
+// TODO: Do we want to allow partial writes? i.e., Skip a failed metric, but
+// try to write all metrics?
```

Comment:
It's just one big write op, isn't it? It's Graphite's responsibility what to do with that write op not completing.

Or are you referring to something similar to https://github.com/prometheus/client_golang/blob/master/prometheus/promhttp/http.go#L142-L161 ? In that case, we have to decide above after L121, in a similar fashion. Could be an option in Config, i.e. `PushPartial` or `ContinueOnError` or something.